tokio/src/io/async_read.rs

```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
+    ///
+    ///     // read the value into a String. We could use any AsyncRead
+    ///     // method here, this is just one example.
+    ///     handle.read_to_string(&mut buffer).await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    fn chain<R>(self, next: R) -> Chain<Self, R>
+    where
+        Self: Sized,
+        R: AsyncRead,
+    {
+        chain(self, next)
+    }
+
+    /// Pulls some bytes from this source into the specified buffer,
+    /// returning how many bytes were read.
+    ///
+    /// Equivalent to:
+    ///
+    /// ```ignore
+    /// async fn read(&mut self, buf: &mut [u8]) -> io::Result<usize>;
+    /// ```
+    ///
+    /// This function does not provide any guarantees about whether it
+    /// completes immediately or asynchronously
+    ///
+    /// If the return value of this method is `Ok(n)`, then it must be
+    /// guaranteed that `0 <= n <= buf.len()`. A nonzero `n` value indicates
+    /// that the buffer `buf` has been filled in with `n` bytes of data from
+    /// this source. If `n` is `0`, then it can indicate one of two
+    /// scenarios:
+    ///
+    /// 1. This reader has reached its "end of file" and will likely no longer
+    ///    be able to produce bytes. Note that this does not mean that the
+    ///    reader will *always* no longer be able to produce bytes.
+    /// 2. The buffer specified was 0 bytes in length.
+    ///
+    /// No guarantees are provided about the contents of `buf` when this
+    /// function is called, implementations cannot rely on any property of the
+    /// contents of `buf` being `true`. It is recommended that *implementations*
+    /// only write data to `buf` instead of reading its contents.
+    ///
+    /// Correspondingly, however, *callers* of this method may not assume
+    /// any guarantees about how the implementation uses `buf`. It is
+    /// possible that the code that's supposed to write to the buffer might
+    /// also read from it. It is your responsibility to make sure that `buf`
+    /// is initialized before calling `read`.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. If an error is returned then it must be
+    /// guaranteed that no bytes were read.
+    ///
+    /// # Examples
+    ///
+    /// [`File`][crate::fs::File]s implement `Read`:
+    ///
+    /// ```no_run
+    /// use tokio::fs::File;
+    /// use tokio::io::{self, AsyncRead};
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> io::Result<()> {
+    ///     let mut f = File::open("foo.txt").await?;
+    ///     let mut buffer = [0; 10];
+    ///
+    ///     // read up to 10 bytes
+    ///     let n = f.read(&mut buffer[..]).await?;
+    ///
+    ///     println!("The bytes: {:?}", &buffer[..n]);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn read<'a>(&'a mut self, buf: &'a mut [u8]) -> Read<'a, Self>
+    where
+        Self: Unpin + Sized,
```

Comment:
```suggestion
        Self: Sized + Unpin,
```

small nit, I prefer this order